### PR TITLE
Move to DataProcessor API

### PR DESCRIPTION
**Pull request by dhpitt**
This PR makes the following changes:

- Creates a `DataProcessor` object that replaces the `torchvision.Transforms` approach to applying transforms inside datasets
- Changes `darcy`and `navier_stokes` datasets to return `DataProcessor` objects
- Moves the `Transform` object from `output_encoder.py` to `transforms.py` and makes it an abstract base class
- Removes `Callback` classes that implement data preprocessing/postprocessing and changes them into `DataPipeline` classes
- a small bug fix and name change from `WandBLoggerCallback` to `BasicLoggerCallback`

### CODE REVIEW

**Review by JeanKossaifi — COMMENTED** [review]
Looks good overall, thanks @dhpitt! I think we still need to polish a little the overall interface.

File: neuralop/training/trainer.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def __init__(self, *,
                  wandb_log=True, 
                  device=None, 
                  amp_autocast=False,
-                 data_processor=None,
+                 data_pipeline=None,
```
[the remark this answers is not in the document]
Comment (dhpitt) [reply neuralop/training/trainer.py:18]: I like your convention of Processor, switching

File: neuralop/training/callbacks.py [hunk 2]
```diff
@@ -225,16 +225,16 @@ def on_val_end(self, *args, **kwargs):
         for c in self.callbacks:
             c.on_val_end(*args, **kwargs)
 
-class SimpleWandBLoggerCallback(Callback):
+class BasicLoggerCallback(Callback):
```
[the remark this answers is not in the document]
Comment (dhpitt) [reply neuralop/training/callbacks.py:228]: The logger is capable of logging to stoud when wandb_log is false

File: neuralop/datasets/transforms.py [hunk 3]
```diff
@@ -17,20 +45,12 @@ class PositionalEmbedding():
...
         self.channel_dim = channel_dim
-        self._grid = None
-
-    def grid(self, data):
-        if self._grid is None:
-            self._grid = get_grid_positional_encoding(data, 
-                                                      grid_boundaries=self.grid_boundaries,
-                                                      channel_dim=self.channel_dim)
-        return self._grid
+        self._grid = None
```
[the remark this answers is not in the document]
Comment (dhpitt) [reply neuralop/datasets/transforms.py]: merged

**Comment by JeanKossaifi** [comment]
Thanks @dhpitt, merging!

**Pull request merged**